### DISCUSSION TITLE
How would a Dockerfile look for keycloak + custom providers?

### DISCUSSION THREAD

**Discussion by dave08** [keycloak/keycloak#10502] [Archive]
In the samples and docs, there's no mention whether the providers need to have build run on them. So when building the dockerfile, do I need to include the custom providers during the build stage. And do I need to copy them to the final image, or does build cache all my code in `/opt/keycloak/lib/quarkus/`, so then I would use the setup in https://www.keycloak.org/server/containers and just add a copy in the first section?

**Comment by stianst** [comment]
Please check the server developer guide

**Comment by dave08** [comment]
I saw: https://www.keycloak.org/server/containers, but there's no mention of PROVIDERS (jars using SPI to customize keycloak, like adding a user provider...) @stianst. Maybe you're referring to something hidden deeper in the docs that I didn't see?

**Comment by stianst** [comment]
Just copy the providers and third-party dependencies into the "providers" directory

**Comment by dave08** [comment]
No need for `kc.sh build`?

**Comment by dave08** [comment]
Actually, it seems like it's needed here: https://www.keycloak.org/docs/latest/server_development/index.html#registering-provider-implementations, but then, do I need to include the jar in the providers folder, or does `kc.sh build` cache all the code in libs/quarkus folder?

**Reply by stianst** [reply]
Yes, sorry you need to both copy the jars into providers, then run kc.sh build afterwards. The latter links the providers and such, so that doesn't have to be done at startup time.

**Reply by dave08** [reply]
>  The latter links the providers and such

Meaning that the jars in providers/ don't have to be copied into the final docker image (sorry, I just want to make sure I'm understanding you properly...)? The guide uses a multi-stage docker build, so whatever is needed in the final image HAS to be copied, and in the guide only the lib/quarkus folder is being copied... which is why I asked this question...

**Reply by stianst** [reply]
Actually not sure about that one. I would think it would need to be copied, but @pedroigor or @DGuhr can confirm.

**Reply by DGuhr** [reply]
Just double checked with the https://github.com/DGuhr/keycloak-quickstarts/tree/latest/extend-account-console example. ATM you'd have to add the jar also to the runtime image, as it is not automatically copied to the lib directory. If not, you get an error like this:

```
ERROR: Failed to start server in (production) mode
ERROR: Failed to open /Users/dguhr/tmp-kc/quarkus/test/keycloak-18.0.0-SNAPSHOT/lib/../providers/keycloak-man-theme.jar
ERROR: /Users/dguhr/tmp-kc/quarkus/test/keycloak-18.0.0-SNAPSHOT/lib/../providers/keycloak-man-theme.jar
```

So no, seems we're not copying the files to the lib directory atm and you have to include the jars in the final docker image.

**Reply by dave08** [reply]
Ok, this might be a good thing to point out in that guide... maybe included in the dockerfile? I'm sure this is not an uncommon thing to do.

**Comment by rafidamr** [comment]
Here is mine, works in 17.0.1

```
FROM quay.io/keycloak/keycloak:17.0.1 as builder

ENV KC_METRICS_ENABLED=true
ENV KC_HEALTH_ENABLED=true
ENV KC_DB=postgres
ENV KC_HTTP_RELATIVE_PATH=/auth

COPY /service-providers-jars /opt/keycloak/providers/
RUN /opt/keycloak/bin/kc.sh build

FROM quay.io/keycloak/keycloak:17.0.1

COPY keycloak-themes/custom /opt/keycloak/themes/custom
COPY --from=builder /opt/keycloak/providers/ /opt/keycloak/providers/
COPY --from=builder /opt/keycloak/lib/quarkus/ /opt/keycloak/lib/quarkus/
```

**Reply by rafidamr** [reply]
the jars are copied to /opt/keycloak/providers/ before `kc.sh build`, then copied again to the final stage
